src/GraphQL/Validation/DocumentValidator.cs

```diff
@@ -80,19 +80,36 @@ public async Task<(IValidationResult validationResult, Variables variables)> Val
                         context.TypeInfo
                     };
 
-                    foreach (var rule in rules)
+                    if (rules is List<IValidationRule> list) //TODO:allocation - optimization for List+Enumerator<IvalidationRule>
```

Comment:
No code changes here - only duplicated branch for List iterator, in future PRs may be removed since design of `BasicVisitor` may be changed.

Reply:
I think this should be eliminated.  There is no need to eliminate one allocation per GraphQL request.  It adds additional complexity with no benefit.

Reply:
I think about it in further PRs - either remove branch or completely rewrite visitors here. Now it helped me to reduce the number of objects in the memory consumption report.

Reply:
Add an issue then.  We should not be adding unnecessarily complex code for the sake of the "memory consumption report"

Reply:
OK. Every allocation (specifically well-know one as for List iterator that can be fixed easily) on green path matters for me. #2886